imap_data_access/file_validation.py

```diff
@@ -0,0 +1,226 @@
+        if not self.is_valid_date(self.enddate):
+            error_message += "Invalid end date format. Please use YYYYMMDD format. \n"
+        if not bool(re.match(r"^v\d{2}-\d{2}$", self.version)):
+            error_message += "Invalid version format. Please use vxx-xx format. \n"
+
+        if self.extension not in imap_data_access.VALID_FILE_EXTENSION or (
+            (self.data_level == "l0" and self.extension != "pkts")
+            or (self.data_level != "l0" and self.extension != "cdf")
+        ):
+            error_message += (
+                "Invalid extension. Extension should be pkts for data "
+                "level l0 and cdf for data level higher than l0 \n"
+            )
+
+        return error_message
+
+    @staticmethod
+    def is_valid_date(input_date: str) -> bool:
+        """Check input date string is in valid format and is correct date
+
+        Parameters
+        ----------
+        input_date : str
+            Date in YYYYMMDD format.
+
+        Returns
+        -------
+        bool
+            Whether date input is valid or not
+        """
+
+        # Validate if it's a real date
+        try:
+            # This checks if date is in YYYYMMDD format.
+            # Sometimes, date is correct but not in the format we want
+            datetime.strptime(input_date, "%Y%m%d")
+            return True
+        except ValueError:
+            return False
+
+    def construct_path(self) -> Path:
+        """Construct valid path from class variables and data_dir.
+
+        If data_dir is not none, it is prepended on the returned path.
+
+        expected return:
+        <data_dir>/mission/instrument/data_level/startdate_month/startdate_day/filename
+
+        Returns
+        -------
+        Path
+            Upload path
+        """
+        upload_path = Path(
+            f"{self.mission}/{self.instrument}/{self.data_level}/"
+            f"{self.startdate[:4]}/{self.startdate[4:6]}/{self.filename}"
+        )
+        if self.data_dir:
+            upload_path = self.data_dir / upload_path
+
+        return upload_path
+
+    @staticmethod
+    def extract_filename_components(filename: str | Path) -> dict:
+        """
+        Extracts all components from filename. Does not validate instrument or level.
+
+        Will return a dictionary with the following keys:
+        { instrument, datalevel, descriptor, startdate, enddate, version, extension }
+
+        If a match is not found, a ValueError will be raised.
+
+        Generally, this method should not be used directly. Instead the class should
+        be used to make a `ScienceFilepath` object.
+
+        Parameters
+        ----------
+        filename : Path or str
+            Path of dependency data.
+
+        Returns
+        -------
+        components : dict
+            Dictionary containing components.
+
+        """
+        pattern = (
+            r"^(?P<mission>imap)_"
```

Comment:
Do you want to take in whatever is passed and not restrict it to imap? Either way is fine I think, so feel free to ignore.
```suggestion
            r"^(?P<mission>[^_]+)_"
```

Reply:
Yeah I wasn't sure which way was best, I just stuck with imap to not change the behavior